File: test-suite-kotlin/src/test/kotlin/io/micronaut/security/docs/securityrule/permitall/PermitAllTest.kt
package io.micronaut.security.docs.securityrule.permitall

import io.micronaut.context.annotation.Property
import io.micronaut.context.annotation.Requires
import io.micronaut.http.HttpRequest
import io.micronaut.http.HttpStatus
import io.micronaut.http.client.HttpClient
import io.micronaut.http.client.annotation.Client
import io.micronaut.http.client.exceptions.HttpClientResponseException
import io.micronaut.security.authentication.AuthenticationFailureReason
import io.micronaut.security.authentication.AuthenticationRequest
import io.micronaut.security.authentication.AuthenticationResponse
import io.micronaut.security.authentication.provider.HttpRequestAuthenticationProvider
import io.micronaut.test.extensions.junit5.annotation.MicronautTest
import jakarta.inject.Singleton
import org.junit.jupiter.api.Assertions.assertDoesNotThrow
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test

@Property(name = "spec.name", value = "PermitAllTest")
@MicronautTest
internal class PermitAllTest {

    @Test
    fun permitAllEndpointCanBeAccessedWithoutAuthentication(@Client("/") httpClient: HttpClient) {
        val client = httpClient.toBlocking()
        assertDoesNotThrow { client.exchange<Any, Any>(HttpRequest.GET("/example/anonymous")) }
    }

    @Test
    fun rolesAllowedEndpointRequiresOneOfTheRoles(@Client("/") httpClient: HttpClient) {
        val client = httpClient.toBlocking()
        var ex = assertThrows(HttpClientResponseException::class.java) {
            client.exchange<Any, Any>(HttpRequest.GET("/example/admin"))
        }
        assertEquals(HttpStatus.UNAUTHORIZED, ex.status)
        ex = assertThrows(HttpClientResponseException::class.java) {
            client.exchange<Any, Any>(HttpRequest.GET<Any>("/example/admin").basicAuth("user", "password"))
        }
        assertEquals(HttpStatus.FORBIDDEN, ex.status)
        assertDoesNotThrow { client.exchange<Any, Any>(HttpRequest.GET<Any>("/example/admin").basicAuth("admin", "password")) }
    }

    @Requires(property = "spec.name", value = "PermitAllTest")
    @Singleton
    internal class AuthenticationProviderUserPassword<B : Any> : HttpRequestAuthenticationProvider<B> {
        override fun authenticate(requestContext: HttpRequest<B>?, authRequest: AuthenticationRequest<String, String>): AuthenticationResponse =
            when (authRequest.identity) {
                "user" -> AuthenticationResponse.success("user")
                "admin" -> AuthenticationResponse.success("admin", listOf("ROLE_ADMIN"))
                else -> AuthenticationResponse.failure(AuthenticationFailureReason.USER_NOT_FOUND)
            }
    }
}

File: test-suite-python/src/test/python/micronaut/security/docs/securityrule/permitall/PermitAllTest.py
from typing import Annotated

from jakarta.inject import Inject, Singleton
from java.util import List
from micronaut.context.annotation import Property, Requires
from micronaut.http import HttpRequest, HttpStatus
from micronaut.http.client import HttpClient
from micronaut.http.client.annotation import Client
from micronaut.http.client.exceptions import HttpClientResponseException
from micronaut.security.authentication import AuthenticationFailureReason, AuthenticationRequest, AuthenticationResponse
from micronaut.security.authentication.provider import HttpRequestAuthenticationProvider
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test


@Property(name="spec.name", value="PermitAllTest")
@MicronautTest
class PermitAllTest:
    httpClient: Annotated[HttpClient, Inject, Client("/")]

    @Test
    def test_permit_all_endpoint_can_be_accessed_without_authentication(self):
        self.httpClient.toBlocking().exchange(HttpRequest.GET("/example/anonymous"))

    @Test
    def test_roles_allowed_endpoint_requires_one_of_the_roles(self):
        client = self.httpClient.toBlocking()
        try:
            client.exchange(HttpRequest.GET("/example/admin"))
            assert False, "HttpClientResponseException expected"
        except HttpClientResponseException as e:
            assert e.getStatus() == HttpStatus.UNAUTHORIZED
        try:
            client.exchange(HttpRequest.GET("/example/admin").basicAuth("user", "password"))
            assert False, "HttpClientResponseException expected"
        except HttpClientResponseException as e:
            assert e.getStatus() == HttpStatus.FORBIDDEN
        client.exchange(HttpRequest.GET("/example/admin").basicAuth("admin", "password"))


@Requires(property="spec.name", value="PermitAllTest")
@Singleton
class AuthenticationProviderUserPassword(HttpRequestAuthenticationProvider):

    def authenticate(self, requestContext: HttpRequest, authRequest: AuthenticationRequest) -> AuthenticationResponse:
        if authRequest.getIdentity() == "user":
            return AuthenticationResponse.success("user")
        if authRequest.getIdentity() == "admin":
            return AuthenticationResponse.success("admin", List.of("ROLE_ADMIN"))
        return AuthenticationResponse.failure(AuthenticationFailureReason.USER_NOT_FOUND)
